pkg/kv/postgres/store.go

```diff
@@ -0,0 +1,239 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/treeverse/lakefs/pkg/kv"
+)
+
+type Driver struct{}
+
+type Store struct {
+	Pool   *pgxpool.Pool
+	Params *Params
+	closed bool
+}
+
+type Entries struct {
+	rows  pgx.Rows
+	entry *kv.Entry
+	err   error
+}
+
+const (
+	DriverName = "postgres"
+
+	defaultTableName = "kv"
+	paramTableName   = "lakefskv_table"
+)
+
+//nolint:gochecknoinits
+func init() {
+	kv.Register(DriverName, &Driver{})
+}
+
+func (d *Driver) Open(ctx context.Context, name string) (kv.Store, error) {
+	// TODO(barak): name format - we need to accept a connection string to the db + params related to db + params related to kv
+	// TODO(barak): handle reuse - open the same kv - reuse the pool
+	config, err := pgxpool.ParseConfig(name)
+	if err != nil {
+		return nil, err
+	}
+	pool, err := pgxpool.ConnectConfig(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		// if we return before store uses the pool, free it
+		if pool != nil {
+			pool.Close()
+		}
+	}()
+
+	// acquire connection and make sure we reach the database
+	conn, err := pool.Acquire(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Release()
+	err = conn.Conn().Ping(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	params := parseStoreConfig(config.ConnConfig.RuntimeParams)
+
+	// migrate database
+	err = migrateDatabase(ctx, conn, params)
+	if err != nil {
+		return nil, err
+	}
+	store := &Store{
+		Pool:   pool,
+		Params: params,
+	}
+	pool = nil
+	return store, nil
+}
+
+type Params struct {
+	TableName string
+}
+
+func parseStoreConfig(runtimeParams map[string]string) *Params {
+	p := &Params{
+		TableName: defaultTableName,
+	}
+	if tableName, ok := runtimeParams[paramTableName]; ok {
+		p.TableName = tableName
+	}
+	return p
+}
+
+func migrateDatabase(ctx context.Context, conn *pgxpool.Conn, params *Params) error {
```

Comment:
Why here and not a DDL package? How would the next migration code look like? How would this migration coexist with pkgs (auth/graveler/etc..) migrations?

Reply:
You are right. I called it migrate because this is the underlying implementation of kv and wanted to support changes in case of internal structure changes - not business level changes.
I'll rename the function, so it will not be confused with the application level migration, which is out of the 'kv' scope.